stream/src/main/java/org/apache/kafka/clients/processor/PTopology.java

```diff
@@ -0,0 +1,176 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.processor;
+
+import org.apache.kafka.common.serialization.Deserializer;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+abstract public class PTopology {
+
+    private List<KafkaProcessor> processors = new ArrayList<>();
+    private Map<String, KafkaSource> sources = new HashMap<>();
+
+    protected final ProcessorProperties properties;
+
+    public PTopology() {
+        this(null);
+    }
+
+    public PTopology(ProcessorProperties properties) {
+        this.properties = properties;
+    }
+
+    public Set<KafkaSource> sources() {
+        Set<KafkaSource> sources = new HashSet<>();
+        for (KafkaSource source : this.sources.values()) {
+            sources.add(source);
+        }
+
+        return sources;
+    }
+
+    public Set<String> topics() {
+        return sources.keySet();
+    }
+
+    public KafkaSource source(String topic) {
+        return sources.get(topic);
+    }
+
+    public Deserializer keyDeser(String topic) {
+        KafkaSource source = sources.get(topic);
+
+        if (source == null)
+            throw new IllegalStateException("The topic " + topic + " is unknown.");
+
+        return source.keyDeserializer;
+    }
+
+    public Deserializer valueDeser(String topic) {
+        KafkaSource source = sources.get(topic);
+
+        if (source == null)
+            throw new IllegalStateException("The topic " + topic + " is unknown.");
+
+        return source.valDeserializer;
+    }
+
+    public final void init(ProcessorContext context) {
+        // init the processors following the DAG ordering
+        // such that parents are always initialized before children
+        Deque<KafkaProcessor> deque = new ArrayDeque<>();
+        for (KafkaProcessor processor : sources.values()) {
+            deque.addLast(processor);
+        }
+
+        while (!deque.isEmpty()) {
+            KafkaProcessor processor = deque.pollFirst();
+
+            boolean parentsInitialized = true;
+            for (KafkaProcessor parent : (List<KafkaProcessor>) processor.parents()) {
+                if (!parent.initialized) {
+                    parentsInitialized = false;
+                    break;
+                }
+            }
+
+            if (parentsInitialized && !processor.initialized) {
+                processor.init(context);
+                processor.initialized = true;
+
+                for (KafkaProcessor child : (List<KafkaProcessor>) processor.children()) {
+                    deque.addLast(child);
+                }
+            }
+        }
```

Comment:
This method appears to completely ignore sources that start out as non-initialized and that contain parents. Is it possible for sources to have parents, and if so is it guaranteed that those parents are already initialized?